book/src/stake-delegation-and-rewards.md

```diff
@@ -0,0 +1,127 @@
+# Stake Delegation and Reward
+
+This proposal covers the current architecture and proposed changes to how stakes
+are delegated to validators.  This design focuses on the software architecture
+for the actual on-chain programs.  Incentives for staking is covered in [staking
+rewardsd](staking-rewards.md).
+
+The main problem this proposal is solving is to allow many delegated stakes to
+generate rewards with a single validator vote without permission from the
+validator.
+
+## Terminology
+
+* VoteState - Instance of the vote program.  This program keeps track of
+validator votes.
+
+* RewardState - Instance of the reward state program.  This program pays out
+rewards for votes.
+
+## Current Architecture
+
+VoteState contains the following state information:
+
+* votes - The submitted votes.
+
+* `delegate_id` - The delegated stake identity.  This identity can submit votes
+as well.
+
+* `credits` - The amount of unclaimed rewards.
+
+* `root_slot` - The last slot to reach the full lockout commitment necessary for
+rewards.
+
+Reward program is stateless, and pays out the reward when that reward is
+claimed.  Claim the reward requires a transaction that includes the following
+instructions:
+
+1. RewardsInstruction::RedeemVoteCredits
+
+2. VoteInstruction::ClearCredits
+
+The payout of the rewards transfers lamports from the rewards program to the
+VoteState address. It ensures that VoteState clears its credits in the next
+instruction.
+
+### VoteInstruction::DelegateStake(Pubkey)
+
+This instruction assigns the delegate\_id to the supplied Pubkey.  The
+delegation allows the delegate\_id to submit votes on behalf of the vote
+program.
+
+## Problems
+
+A validator that has been delegated N stakes needs to submit N votes, and the
```

Comment:
Is there no way to have all the accounts sign a vote the delegate generates?

Also in the proposed solution, there's no node or hardware assigned as the delagte right? How should stakes for a node be calculated?

Reply:
@sagar-solana 

1. That is another approach, but that requires delegates to be active and running their own vote signing hardware.  It has some benefits though.
2. The RewardState instance has the identity of the VoteState, so for any given VoteSate, the total delegated stake can be calculated from the accounts that have it as the `vote_state_id`

Reply:
1. @garious had suggested an "authorize vote signer" instruction for vote program. In theory if a vote signer is being used, the delegate (node) could send a vote to the signer for each assigned vote account and the signer could decide whether or it wants to sign that vote. This complicates the definition of stake for a node a bit though. 

2. It looks like you removed delegate_id from the VoteState in the proposal section. RewardsState has vote_state_id which ties it to a vote_account. But neither seems to have the ability to delegate to a node for voting and leader rotation. How does that work? I could totally be missing something here..

Reply:
1. Yep, but that doesn't solve the problem of many many more token holders with stake that are interested in staking vs the number of individual validator nodes running.

2. leader rotation is outside the scope.  But it would be just a field in the VoteState.

Reply:
1. I know @pgarg66 or @rob-solana can help with this. We vaguely discussed multi-sig on a vote. Making a validator vote just once and have each delegated account's vote signer (could be different or the same) decide whether or not it wants to sign a vote. That could enable a one-to-many validator-stakers relation. 

2. Understood. :)

Reply:
@sagar-solana but you still need to submit N transactions to the network, and each signer needs to be active.  So what is really great about that approach is that it adds an extra level of security.  The vote signer can protect itself from getting slashed.  But it requires high availability, and humans are super lazy about availability.

Reply:
Ah you're totally right, still needs N txs. Okay some advantages but doesn't reduce network burden.

Reply:
@aeyakovenko, 

1) Why do we need to submit N votes for N delegated stakes, instead of just sending one vote and mapping that vote to all the VoteStates who have delegated to that validator? 

2) Are we using one epoch warm up/cool down periods for staking/changing delegates?

3) Will it be the responsibility of individual nodes to make sure they don't get slashed when they transition delegates? (aka wait a certain cooldown period for lockouts to expire, or make sure that the delegate is on a compatible fork?)

Reply:
@carllin 
1. to implement 1, the bank would need to pull up all the accounts for delegates stakes on every time it executes a vote transaction and keep that set consistent for every vote while allow for votes to be processed concurrently.
2. outside of this proposal
3. passive delegates rely on the validator to do the right thing.  so they are on the hook for a misbehaved validator.